core/src/main/java/com/linecorp/armeria/server/EmptyContentDecodedHttpRequest.java

```diff
@@ -0,0 +1,163 @@
+
+    @Nullable
+    private HttpResponse response;
+    private boolean isResponseAborted;
+
+    EmptyContentDecodedHttpRequest(EventLoop eventLoop, int id, int streamId, RequestHeaders headers,
+                                   boolean keepAlive) {
+        delegate = HttpRequest.of(headers);
+        this.eventLoop = eventLoop;
+        this.id = id;
+        this.streamId = streamId;
+        this.keepAlive = keepAlive;
+    }
+
+    @Override
+    public void init(ServiceRequestContext ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public int id() {
+        return id;
+    }
+
+    @Override
+    public int streamId() {
+        return streamId;
+    }
+
+    /**
+     * Returns whether to keep the connection alive after this request is handled.
+     */
+    @Override
+    public boolean isKeepAlive() {
+        return keepAlive;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return delegate.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return delegate.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpObject> subscriber, EventExecutor executor) {
+        delegate.subscribe(subscriber, executor);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpObject> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        delegate.subscribe(subscriber, executor, options);
+    }
+
+    @Override
+    public EventLoop defaultSubscriberExecutor() {
+        return eventLoop;
+    }
+
+    @Override
+    public void abort() {
+        delegate.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        delegate.abort(cause);
+    }
+
+    @Override
+    public RequestHeaders headers() {
+        return delegate.headers();
+    }
+
+    @Override
+    public void close() {}
+
+    @Override
+    public void close(Throwable cause) {}
+
+    @Override
+    public void setResponse(HttpResponse response) {
+        if (isResponseAborted) {
+            // This means that we already tried to close the request, so abort the response immediately.
+            if (!response.isComplete()) {
+                response.abort();
+            }
+        } else {
+            this.response = response;
+        }
+    }
+
+    @Override
+    public void abortResponse(Throwable cause, boolean cancel) {
+        isResponseAborted = true;
+
+        // Make sure to invoke the ServiceRequestContext.whenRequestCancelling() and whenRequestCancelled()
+        // by cancelling a request
+        if (cancel && ctx != null) {
+            ctx.cancel(cause);
+        }
+
+        if (response != null && !response.isComplete()) {
+            response.abort(cause);
+        }
```

Comment:
I saw some duplicate logic but I guess this is the best we can do for now. 😄

Reply:
Because of the class hierarchy, I copy and paste the logic. 🙈 For better abstraction, we need to compose `DefaultStreamMessage` and remove it from the super class.